### DISCUSSION TITLE
Input changes: phase 2

### DISCUSSION THREAD

**Discussion by cuavas** [mamedev/discussions#26] [General]
Hi everyone!  I was hoping to have this stuff in shape a bit earlier, but I got sidetracked with the rendering stuff.  On the bright side, BGFX should be in better shape for MAME 0.252 as well.  Anyway, as of mamedev/mame@d4589e0b29d6085072ff635d87c3d95c21813a58, MAME should be smarter about its default input assignments for game controllers.  This changes quite a few things.

I’d appreciate if people with exotic controllers (e.g. @darkmos, @MrDo-Arcade and @sairuk) could build from the latest source and try things out.

# New UI inputs

UI input have changed.
* The input previously called “Config Menu” is now called “Show/Hide Menu” to better reflect what it actually does.
* There’s a new “UI Help” input for showing context-sensitive help (currently only used by the Analog Input Adjustments menu).
* There’s a new “UI Back” input that takes over some of the functions that were previously part of “UI Cancel”.  They currently both default to using the Escape key, so your muscle memory will still work.

The third change is the most significant.  This means that a game controller button can be assigned to “UI Back” to allow menus to be closed easily without having to avoid that button while playing.

The only real drawback is that assigning combinations involving the “UI Cancel” key is annoying (you effectively need to clear the “UI Cancel” assignment while you set up the other assignments), and there’s potentially a need to do that now.

# More functionality in SDL builds

Windows SDL builds now include DirectInput, XInput and XAudio2 modules.  This can give faster turnaround for developers as more modules can be tested in a single build.  It’s probably not all that exciting for users, unless you have some use case where you want to mix and match different modules.

# Better input behaviour

The threshold for treating an axis as “on” when used as a switch is not a separate setting from the dead zone.  This means the default dead zone can be made smaller, giving more useful range for analog sticks, triggers and pedals, and you can set the dead zone to zero without having UI inputs assigned to analog axes going nuts.

More modules honour the `background_input` setting.  DirectInput, XInput, and SDL game controller and joystick modules all honour the setting if that’s what you want to do.  Various issues with DirectInput have been fixed, too.

# Smarter default assignments for game controllers

This is obviously the big thing.  Default input assignments now take the type of controller and available controls into consideration, to varying extents depending on the module.  Default assignments should be more appropriate for your controllers, and it’s now possible to navigate more of the internal UI with a game controller by default.  Most controllers should now have default assignments for UI Back and UI Clear, which are rather important controls.

## XInput (xinput, winhybrid)

Most controllers should work well:
* Gamepads should have secondary UI navigation controls on the shoulder buttons, triggers and second analog stick.  Dual analog sticks should work correctly for dual joystick games (e.g, Smash TV, Libble Rabble and Crazy Climber).
* Arcade sticks should have fairly intuitive default assignments, although some of them don’t correctly report the absence of some controls.
* Flight sticks should have secondary UI navigation controls on the hat switch, throttle, rudder, and/or D-pad if present.  This is based on documentation and assumptions as I don’t actually have an XInput flight stick to test with.
* Guitars should be less annoying.  The accelerometers aren’t assigned to inputs by default, you can navigate menus using the D-pad and strum bar without having to hold the guitar perfectly level.  The whammy bar should work as the accelerator pedal in driving games, and for switching panes on the system and software selection menus.  If the neck slider is present, it’s used for steering and paddle controls by default.
* Drum kits should just show up with D-pad and button assignments.  The bass drum pedal should switch panes on the system and software selection menus.
* The turntable controller should use the turntable for most horizontal game controls, and the effect knob for most vertical game controls.  The turntable should move up and down menus, and the effect knob should adjust settings.
* The Rock Band keyboard should be set up as a mahjong/hanafuda controller by default (white keys to discard tiles, black keys for other actions).

XInput driving controls might have bad default assignments, I haven’t really thought about them, and I don’t have any to test with.

## DirectInput (dinput, winhybrid)

Gamepads, flight sticks and driving controls should all have reasonable default assignments.  For flight sticks and driving controls, I’m working with limited documentation and assumptions.

There’s no semantic information about buttons, so it can’t try to be as clever as XInput.  It’s probably not going to work too well if your first controller is a limited or auxiliary controller (e.g. a throttle quadrant or pedal unit).

## SDL Game Controller (sdlgame)

Triggers should be assigned to pedals by default, and dual analog sticks should be correctly assigned for dual joystick games (e.g, Smash TV, Libble Rabble and Crazy Climber).  Shoulder buttons, triggers and the second analog stick are used as additional navigation controls.  The home/logo button shows or hides the menu.

Exotic controllers aren’t catered for with this module at the moment.

## SDL Joystick (sdljoy)

As this module doesn’t have access to semantic information about the controls, this module follows MAME’s traditional approach to assigning controls to inputs.  The second analog stick on a gamepad or the throttle and rudder on a flight stick can be used to switch between panes on the system and software selection menus.

**Comment by darkmos** [comment]
Just a heads up I've been busy this week-end but I'm not forgetting you, I'll have a first look this evening after work with the gamepads/arcade stick/flight stick and I'll probably check the instruments later this week as I need to fetch them from storage.

**Reply by darkmos** [reply]
mame -version: 0.252 (mame0252-50-g14d00bf2f75)

**Xbox360 X-plorer Wired Guitar**
```
Skipping DirectInput for XInput compatible joystick 0x6067bf8c68.
XInput: Configuring player 1 type 0x01 (game controller) sub type 0x07 (alternate guitar).
XInput: Switch capabilities A=1 B=1 X=1 Y=1 LB=1 RB=1 LSB=0 RSB=0 Start=1 Back=1 Up=1 Down=1 Left=1 Right=1.
XInput: Axis capabilities LT=0xFF (8-bit) RT=0xFF (8-bit) LSX=0x0000 (0-bit) LSY=0x0000 (0-bit) RSX=0xFFC0 (10-bit) RSY=0xFFC0 (10-bit).
Input: Adding joystick #1: XInput Player 1 (device id: XInput Player 1)
```
Behavior as described e.g. whammy bar switches zones/panes in the menu

**Xbox360 Madcatz Portable Drums for Rockband**
```
Skipping DirectInput for XInput compatible joystick 0xf2f15f8cb8.
XInput: Configuring player 1 type 0x01 (game controller) sub type 0x08 (drum kit).
XInput: Switch capabilities A=1 B=1 X=1 Y=1 LB=1 RB=0 LSB=0 RSB=0 Start=1 Back=1 Up=1 Down=1 Left=1 Right=1.
XInput: Axis capabilities LT=0x00 (0-bit) RT=0x00 (0-bit) LSX=0x0000 (0-bit) LSY=0x0000 (0-bit) RSX=0x0000 (0-bit) RSY=0x0000 (0-bit).
Input: Adding joystick #1: XInput Player 1 (device id: XInput Player 1)
```
Behavior as described e.g. bass drum pedal switches zones/panes in the menu

**DJ Hero 2 Turntable**
```
Skipping DirectInput for XInput compatible joystick 0xdb11f91d8.
XInput: Configuring player 1 type 0x01 (game controller) sub type 0x17 (unsupported).
XInput: Switch capabilities A=1 B=1 X=1 Y=1 LB=1 RB=1 LSB=1 RSB=1 Start=1 Back=1 Up=1 Down=1 Left=1 Right=1.
XInput: Axis capabilities LT=0xFF (8-bit) RT=0xFF (8-bit) LSX=0x1430 (0-bit, invalid) LSY=0x1705 (0-bit, invalid) RSX=0x0001 (0-bit, invalid) RSY=0xFFC0 (10-bit).
XInput: Detected DJ Hero turntable controller.
Input: Adding joystick #1: XInput Player 1 (device id: XInput Player 1)
```
Behavior as described e.g. was able to use the turntable to steer the car in outrun

**Xbox Serie X wireless controller**
```
Skipping DirectInput for XInput compatible joystick 0x1b00df8ee8.
XInput: Configuring player 1 type 0x01 (game controller) sub type 0x01 (gamepad).
XInput: Switch capabilities A=1 B=1 X=1 Y=1 LB=1 RB=1 LSB=1 RSB=1 Start=1 Back=1 Up=1 Down=1 Left=1 Right=1.
XInput: Axis capabilities LT=0xFF (8-bit) RT=0xFF (8-bit) LSX=0xFFC0 (10-bit) LSY=0xFFC0 (10-bit) RSX=0xFFC0 (10-bit) RSY=0xFFC0 (10-bit).
Input: Adding joystick #1: XInput Player 1 (device id: XInput Player 1)
```
Behavior as described e.g. LB/RB change zone/panes in the menu, right stick act as PG up/PG Down in the gamelist, dual joystick works by default in Smash TV and Little Rabble

**Reply by darkmos** [reply]
@cuavas do you have a specific test scenario or peripheral in mind? As far as I can see things works as expected.

**Reply by cuavas** [reply]
That’s basically it – things should work as you’d expect.  If there’s something that feels wrong or unintuitive with one or more of your controllers, I want to know.  The idea is that MAME should be more usable out-of-the-box with a controller without the need to do extra setup.

**Reply by darkmos** [reply]
**Xbox Serie X wireless controller** with sdlgame
```
Game Controller:   ...  mapping [0300509d5e040000120b000000007200,*,a:b0,b:b1,x:b2,y:b3,back:b6,guide:b10,start:b7,leftstick:b8,rightstick:b9,leftshoulder:b4,rightshoulder:b5,dpup:h0.1,dpdown:h0.4,dpleft:h0.8,dpright:h0.2,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:a4,righttrigger:a5,crc:9d50,platform:Windows]
Input: Adding joystick #1: Xbox Series X Controller (device id: 0300509d5e040000120b000000007200)
Game Controller:   ...  Xbox One type
```
Launched Outrun and game automatically uses RT for accelerator and LT for brake in outrun, guide button shows the menu

**Reply by darkmos** [reply]
> That’s basically it – things should work as you’d expect. If there’s something that feels wrong or unintuitive with one or more of your controllers, I want to know. The idea is that MAME should be more usable out-of-the-box with a controller without the need to do extra setup.

Ok I'll keep switching through with my gamepads when using mame and if I see something strange I'll react here otherwise you can assume it's all good. But for sure it's a great improvement in usability :)

**Comment by sairuk** [comment]
**8bitdo SF30 xinput wired**
```
Game Controller: Start initialization
Game Controller: Xbox 360 Controller [GUID 030003f05e0400008e02000014016800] Vendor ID 045E, Product ID 028E, Revision 0114, Serial <nullptr>
Game Controller:   ...  mapping [030003f05e0400008e02000014016800,*,a:b0,b:b1,back:b4,dpdown:b12,dpleft:b13,dpright:b14,dpup:b11,guide:b5,leftshoulder:b9,leftstick:b7,lefttrigger:a4,leftx:a0,lefty:a1,rightshoulder:b10,rightstick:b8,righttrigger:a5,rightx:a2,righty:a3,start:b6,x:b2,y:b3,crc:f003,platform:Linux]
Input: Adding joystick #1: Xbox 360 Controller (device id: 030003f05e0400008e02000014016800)
Game Controller:   ...  Xbox 360 type
Game Controller: End initialization
```
xinput mode A/B are swapped in regards to physical vs mapped

as such B operates as Select vs A as listed in the UI input options

I do not recall this operating this way previously, but can rebuild and test if needed.

**8bitdo m30 xinput wired**
```
Game Controller: Start initialization
Game Controller: Xbox 360 Controller [GUID 030003f05e0400008e02000014016800] Vendor ID 045E, Product ID 028E, Revision 0114, Serial <nullptr>
Game Controller:   ...  mapping [030003f05e0400008e02000014016800,*,a:b0,b:b1,back:b4,dpdown:b12,dpleft:b13,dpright:b14,dpup:b11,guide:b5,leftshoulder:b9,leftstick:b7,lefttrigger:a4,leftx:a0,lefty:a1,rightshoulder:b10,rightstick:b8,righttrigger:a5,rightx:a2,righty:a3,start:b6,x:b2,y:b3,crc:f003,platform:Linux]
Input: Adding joystick #1: Xbox 360 Controller (device id: 030003f05e0400008e02000014016800)
Game Controller:   ...  Xbox 360 type
Game Controller: End initialization
```

UI maps seem OK
Left shoulder maps to LB
Right shoulder maps to LT
C button maps to RT

recall this was the same previously, needs a custom map

**8bitdo pro 2 xinput wired**
```
Game Controller: Start initialization
Game Controller: Xbox 360 Controller [GUID 030003f05e0400008e02000014016800] Vendor ID 045E, Product ID 028E, Revision 0114, Serial 000000000003
Game Controller:   ...  mapping [030003f05e0400008e02000014016800,*,a:b0,b:b1,back:b4,dpdown:b12,dpleft:b13,dpright:b14,dpup:b11,guide:b5,leftshoulder:b9,leftstick:b7,lefttrigger:a4,leftx:a0,lefty:a1,rightshoulder:b10,rightstick:b8,righttrigger:a5,rightx:a2,righty:a3,start:b6,x:b2,y:b3,crc:f003,platform:Linux]
Input: Adding joystick #1: Xbox 360 Controller (device id: 030003f05e0400008e02000014016800)
Game Controller:   ...  Xbox 360 type
Game Controller: End initialization
```

Same situation as the SF30
